### DISCUSSION TITLE
Using Cucumber with an existing Vitest setup

### DISCUSSION THREAD

**Discussion by charlieforward9** [vitest-dev/vitest#5515] [Q&A] [answered]
I have been looking into using Cucumber to test integrations while using Vitest test the unit and component lifecycles. In practice, I am curious about a few things that seem incomplete about using these tools together in a seamless manner.

1. Writing Cucumber tests is distinct from Vitest tests. Given this, I imagine the coverages are not merged with the Vitest coverage report. I have not tested this myself, but I am curious if this is this possible to do? Particularly with an actions tool like [vitest-coverage-report-action](https://github.com/davelosert/vitest-coverage-report-action)?

2. Using [vitest-cucumber](https://github.com/amiceli/vitest-cucumber) removes the 'quick-run' button that shows to the left of the line number on the `describe`, `it`, or `test` lines in my test files, since it replaces these APIs with Cucumber-specific `describeFeature`, etc method. I opened [#62](https://github.com/amiceli/vitest-cucumber/issues/62) to continue this conversation, but would like to know how likely it would be to see this come to fruition sometime soon.

3. Using [vitest-cucumber-plugin](https://github.com/samuel-ziegler/vitest-cucumber-plugin), does not seem any easier. Especially as it has not been contributed to in 9 months. Additionally, I am trying to use React Testing Library to render the component I am testing, which does not look to have any documentation/examples.

Given these points, it seems that my best bet is to keep Cucumber and Vitest separate. However, I'd like to hear back from some contributors who may be able to help me get a deeper understanding of this use case.

**Comment by sheremet-va** [comment, marked as answer]
1. Not sure why it would be different. Coverage is just running files - if you import a file and run its methods, the coverage would be fine
2. Try using the pre-release version of VS Code extension. It should support the quick-run buttons. It also support coverage by the way
3. Cannot help you here, unfortunately

**Reply by charlieforward9** [reply]
Ok, it looks like with [vitest-cucumber](https://github.com/amiceli/vitest-cucumber) and the pre-release Vitest VSCode Extension, I was able to address 1 and 2.

Thank you @sheremet-va 

Small note on the UI here, the quick run buttons are not where they should be. I have the feature starting at line 17 and the Scenario at line 26 but the quick run button is at line 30, randomly next to the And keyword. 

The tests show up as expected in the Test Explorer.

This does not seem to be expected behavior for the quick run button.

![image](https://github.com/vitest-dev/vitest/assets/62311337/330500f8-3cd9-4c5a-b83f-167eb1c907dc)

**Reply by sheremet-va** [reply]
The position depends on the stack trace - in your case I would assume most of the tests are in `node_modules/vitest-cucumber` (you can click on the last button in the tree view to open the source

**Reply by sheremet-va** [reply]
The position is correct, by the way. `vitest-cucumber` uses `test` under the hood for `And`

**Reply by charlieforward9** [reply]
I am having issues with the test output not showing the correct output.

It would also be nice to see the breakdown of quick-run buttons on the lines similar to how it is broken down in the Test Explorer. This integration could use a bit of work, I will try to log the issues in the relevant area.

**Reply by amiceli** [reply]
I'm maintainer of vitest-cucumber, maybe I can help.
vitest-cucumber use `test` in each scenario steps like `And`, `Given` etc with `test.each`.
And `describeFeature`, `Scenario` and `ScenarioOutline` use `describe`.

@sheremet-va I'm not an expert, you speak about stack-race maybe I can improve this in vitest-cucumber project ?

**Reply by sheremet-va** [reply]
>  I'm not an expert, you speak about stack-race maybe I can improve this in vitest-cucumber project ?

It should just work out of the box. Vitest has a location property that resolves to the location of the call in a test file, so `And`, `Given`, etc will be displayed as tests, and `describeFeature`, `Scenario` and `ScenarioOutline` are displayed as suites.

I increased the stack limit so it should hopefully display icons correctly for all of them now.

**Reply by charlieforward9** [reply]
Do I have to update something? I am still not seeing the icons correctly.
